File: Advanced_Robotics_Course/Chapter10/Lesson1/Chapter10_Lesson1.py
import numpy as np
import open3d as o3d

def rigid_transform_svd(P, Q):
    """
    Solve for R, t that minimize sum ||R p_i + t - q_i||^2
    P, Q: (N, 3) numpy arrays of corresponding points.
    """
    assert P.shape == Q.shape
    N = P.shape[0]

    # Compute centroids
    p_bar = P.mean(axis=0)
    q_bar = Q.mean(axis=0)

    # Center the points
    P_centered = P - p_bar
    Q_centered = Q - q_bar

    # Cross-covariance matrix H
    H = P_centered.T @ Q_centered

    # SVD of H
    U, S, Vt = np.linalg.svd(H)
    R = Vt.T @ U.T

    # Handle possible reflection
    if np.linalg.det(R) < 0:
        Vt[2, :] *= -1.0
        R = Vt.T @ U.T

    t = q_bar - R @ p_bar
    return R, t

# Example: using Open3D for ICP with point-to-point metric
source = o3d.io.read_point_cloud("model_cloud.pcd")
target = o3d.io.read_point_cloud("scene_cloud.pcd")

# Downsample for speed
source_ds = source.voxel_down_sample(voxel_size=0.005)
target_ds = target.voxel_down_sample(voxel_size=0.005)

# Initial guess (identity)
init_transform = np.eye(4)

threshold = 0.02  # max correspondence distance in meters
icp_result = o3d.pipelines.registration.registration_icp(
    source_ds,
    target_ds,
    threshold,
    init_transform,
    o3d.pipelines.registration.TransformationEstimationPointToPoint()
)

print("ICP fitness:", icp_result.fitness)
print("ICP inlier_rmse:", icp_result.inlier_rmse)
print("Estimated transform:\n", icp_result.transformation)
      

File: Tutorials/Codes/Advanced_Robotics_Course/Chapter10/Lesson1/Chapter10_Lesson1.m
% Load point clouds
model = pcread("model_cloud.pcd");
scene = pcread("scene_cloud.pcd");

% Downsample
voxelSize = 0.005;
modelDS = pcdownsample(model, "gridAverage", voxelSize);
sceneDS = pcdownsample(scene, "gridAverage", voxelSize);

% ICP registration (point-to-point)
[tform, movingReg, rmse] = pcregistericp( ...
    modelDS, sceneDS, ...
    "Metric", "pointToPoint", ...
    "MaxIterations", 50, ...
    "Tolerance", [1e-6, 1e-6]);

R_icp = tform.T(1:3, 1:3);
t_icp = tform.T(1:3, 4);

disp("ICP rotation:");
disp(R_icp);
disp("ICP translation:");
disp(t_icp);

% SVD-based rigid transform from explicit correspondences
function [R, t] = rigidTransformSVD(P, Q)
% P, Q: N-by-3 matrices of corresponding points
assert(all(size(P) == size(Q)), "P and Q must have same size");
N = size(P, 1);

p_bar = mean(P, 1);
q_bar = mean(Q, 1);

P_centered = P - p_bar;
Q_centered = Q - q_bar;

H = P_centered' * Q_centered;
[U, S, V] = svd(H);

R = V * U';
if det(R) < 0
    V(:, 3) = -V(:, 3);
    R = V * U';
end

t = q_bar' - R * p_bar';
end
      
